### PR TITLE
feat: Recurring Additional Salary

### PR DESCRIPTION
**Pull request by Anurag810**
image

### CODE REVIEW

**Comment by sammish** [comment]
@Anurag810  Please add GIF videos together.Because i am also working HR module I want to know coming changes.

**Comment by Anurag810** [comment]
Docs: https://github.com/frappe/erpnext_com/pull/656

File: erpnext/hr/doctype/additional_salary/additional_salary.py [hunk 1]
```diff
@@ -21,10 +21,23 @@ def validate(self):
...
 	def validate_dates(self):
- 		date_of_joining, relieving_date = frappe.db.get_value("Employee", self.employee,
+		date_of_joining, relieving_date = frappe.db.get_value("Employee", self.employee,
 			["date_of_joining", "relieving_date"])
- 		if date_of_joining and getdate(self.payroll_date) < getdate(date_of_joining):
- 			frappe.throw(_("Payroll date can not be less than employee's joining date"))
+
+		if not self.is_recurring and not self.payroll_date:
+			frappe.msgprint(_("Please enter Payroll Date."), indicator='blue', raise_exception=1)
+		if self.is_recurring and not self.from_date and not self.to_date:
```
Comment (nabinhait) [inline erpnext/hr/doctype/additional_salary/additional_salary.py]: Use "Mandatory Depends On" property for such validation.

"To Date" should be optional, many times user will not know the End Date, "Allow on Submit" should be enabled.

File: erpnext/hr/doctype/salary_slip/salary_slip.py [hunk 2]
```diff
@@ -402,6 +402,7 @@ def add_additional_salary_components(self):
 		if additional_components:
 			for additional_component in additional_components:
 				amount = additional_component.amount
+				print("-------------[>>>]", amount)
```
Comment (nabinhait) [inline erpnext/hr/doctype/salary_slip/salary_slip.py]: remove print

File: erpnext/hr/doctype/additional_salary/additional_salary.py [hunk 3]
```diff
@@ -45,7 +58,11 @@ def get_additional_salary_component(employee, start_date, end_date):
 		from `tabAdditional Salary`
 		where employee=%(employee)s
 			and docstatus = 1
-			and payroll_date between %(from_date)s and %(to_date)s
+			and (payroll_date between %(from_date)s and %(to_date)s)
+			or (
+				(from_date between %(from_date)s and %(to_date)s)
+				or(to_date between %(from_date)s and %(to_date)s)
```
Comment (nabinhait) [inline erpnext/hr/doctype/additional_salary/additional_salary.py]: it will give error if we make to_date optional

**Review by nabinhait — CHANGES_REQUESTED** [review]
- Need to fix `update_salary_slip_in_additional_salary` function in salary_slip.py
- Add tests

File: erpnext/hr/doctype/additional_salary/additional_salary.py (hunk 1 above)
Comment (Anurag810) [reply erpnext/hr/doctype/additional_salary/additional_salary.py]: done

File: erpnext/hr/doctype/salary_slip/salary_slip.py (hunk 2 above)
Comment (Anurag810) [reply erpnext/hr/doctype/salary_slip/salary_slip.py]: done

File: erpnext/hr/doctype/additional_salary/additional_salary.py (hunk 3 above)
Comment (Anurag810) [reply erpnext/hr/doctype/additional_salary/additional_salary.py]: made both mandatory if ``` is_recurring = 1 ```

File: erpnext/hr/doctype/additional_salary/additional_salary.py [hunk 4]
```diff
@@ -45,7 +54,11 @@ def get_additional_salary_component(employee, start_date, end_date):
...
```
Comment (nabinhait) [inline erpnext/hr/doctype/additional_salary/additional_salary.py]:
```suggestion
				from_date <= %(to_date)s and to_date >= %(to_date)s
```

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/hr/doctype/additional_salary/additional_salary.py (hunk 4 above)
Comment (Anurag810) [reply erpnext/hr/doctype/additional_salary/additional_salary.py]: thanks, Nabin. done

**Comment by Anurag810** [comment]
@nabinhait please review and merge

**Comment by Anurag810** [comment]
open another PR

**Pull request closed without merging**